Fix: fee recovery in the “donation summary” with wrong value

glaubersilva opened the pull request:
<!-- Make sure to prefix the title with one of New:, Fix:, Changed:, or Security: -->

<!-- Indicate the issue(s) resolved by this PR. -->

Resolves #6536

## Description

<!-- Summarize the related issue, explain HOW this PR solves the problem, and WHY you made the choices you made. -->

The donation summary is displaying the wrong fee recovery amount for currencies that uses a comma as a decimal separator. 

~~This PR solves the problem by adding extra param (with "true" as the default value) to the `format_amount()` function that allows us to set if the amount should be normalized or not because when it's called inside the `handleFees()` function the value passed through is already normalized.~~

~~So, when the value is already normalized, we need to set this extra param to "false" to prevent the double normalization process - because when it happens, a wrong value is generated.~~

This PR solves the problem by getting the formatted value directly from the DOM instead of retrieving it from data attributes and using the `format_amount()` function to normalize these values - in fact, this function is no longer useful and has been removed.

Another problem solved by this PR is that now the fee amount gets updated even when the `GiveWP > Settings > Fee Recovery > Fee Opt-In > Forced Opt-in` is enabled - this option removes the checkbox observed on the frontend and it also was preventing the fee amount update. 

Now the frontend no more observes the checkbox to handle the fee amount changes, instead that the `handleFees()` function is triggered every time that the `.give-final-total-amount` element gets updated. This way, some old codes responsible for updating the fee amount were removed from some files because they are no more necessary.

## Affects

<!-- Mention any existing functionality affected by this PR to help inform the reviewer(s). -->

## Visuals

<!-- Include screenshots or video to better communicate your changes. -->

Before fix: https://youtu.be/PcwG0vrvybU

After fix: https://youtu.be/jDG-YHV1ru8

## Testing Instructions

<!-- Help others test your PR as efficiently as possible. -->

1. Install and active the Fee Recovery Add-on
2. Setup a multi-step or classic form template
3. Use a currency that uses a comma as a decimal separator like Euro (€)
4. Make a test donation and tick the "cover transaction cost" option
5. Before finishing the donation, the fee value in the “donation summary” should accurately reflect the fee recovery amount
6. Also make a second test donation with the `GiveWP > Settings > Fee Recovery > Fee Opt-In > Forced Opt-in` enabled

## Pre-review Checklist

<!-- Complete tasks prior to requesting a review. Add to this list, but do not remove the base items. -->

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [ ] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [ ] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

ravinderk commented on src/DonationSummary/resources/js/summary.js:
```diff
@@ -205,12 +205,15 @@ window.GiveDonationSummary = {
+        if (normalize) {
+            amount = amount
+                .replace(Give.form.fn.getInfo('thousands_separator', $form), '')
+                .replace(Give.form.fn.getInfo('decimal_separator', $form), '.');
```
@glaubersilva Can you confirm that this logic is compatible with the currency that has `,` as decimal separator like `EUR`

glaubersilva replied:
I did a double check and could notice that the "Donation Total" amount was getting wrong when the value was higher than 999, so I update the normalize amounts logic and removed the extra param added previously.

ravinderk commented on src/DonationSummary/resources/js/summary.js:
```diff
@@ -205,12 +205,13 @@ window.GiveDonationSummary = {
+     * @param {boolean} normalize
      */
     format_amount: function (amount, $form) {
         // Normalize amounts to JS number format
```
@glaubersilva Can you use `Give.fn.unFormatCurrency` and `Give.fn.formatCurrency`. These functions encapsulate current formatting setting logic for a donation form and provide accurate results with add-ons.

ravinderk replied:
@glaubersilva keep in mind that currencies can have ` ` [space], `,`, `.` as thousand separators which can cause unexpected results.

ravinderk replied:
fixed: https://github.com/impress-org/givewp/commit/6c17283816f75b0f160591b112c7ac8fd9212450

ravinderk reviewed (changes requested):
@glaubersilva I noticed another issue. If the fee does not display in the donation summary when the admin forces fee optin to the donor.

<img width="1395" alt="Screenshot 2022-08-18 at 7 27 17 PM" src="https://user-images.githubusercontent.com/1784821/185414065-d982bc9a-dcc9-4d96-a9cd-a4d8eb64ea2f.png">
<img width="724" alt="image" src="https://user-images.githubusercontent.com/1784821/185414247-be99b913-c288-4cf0-847b-cbb9fb45c1e5.png">

glaubersilva commented:
@ravinderk I implemented the requested change and also fixed the format amount args. However, the total donation amount still is getting wrong because of [another issue](https://github.com/impress-org/give-fee-recovery/issues/306) that already has this [other PR](https://github.com/impress-org/give-fee-recovery/pull/307) avaliable.

jonwaldstein commented:
@ravinderk can you please update the PR description to describe the additional changes you made? I see a lot of things were removed and replaced and it's a little hard to understand why those decisions were made by looking at it.

glaubersilva commented:
@jonwaldstein I updated the PR description with more details as @ravinderk asked me, but if it's not clear enough, let me know.

jonwaldstein commented on src/DonationSummary/resources/js/summary.js:
```diff
@@ -1,22 +1,40 @@
+
         GiveDonationSummary.observe('[name="give-amount"]', function (targetNode, $form) {
-            $form.find('[data-tag="amount"]').html(GiveDonationSummary.format_amount(targetNode.value, $form));
+            $form.find('[data-tag="amount"]').html(getFormattedDonationAmount($form));
```
Let's remove this local instance `const getFormattedDonationAmount = this.getFormattedDonationAmount;` in favor of the parent method.

```suggestion
            $form.find('[data-tag="amount"]').html(GiveDonationSummary.getFormattedDonationAmount($form));
```

jonwaldstein commented on src/DonationSummary/resources/js/summary.js:
```diff
@@ -100,50 +118,51 @@ window.GiveDonationSummary = {
-            $form.find('[data-tag="total"]').html(GiveDonationSummary.format_amount(total, $form));
+            $form.find('[data-tag="total"]').html(targetNode.textContent);
+
+            handleFees($form);
```
Same here, we don't need to have a local instance of `const handleFees = this.handleFees;`.

```suggestion
            GiveDonationSummary.handleFees($form);
```

jonwaldstein commented on src/DonationSummary/resources/js/summary.js:
```diff
@@ -100,50 +118,51 @@ window.GiveDonationSummary = {
         const totalAmount = document.querySelector('.give-final-total-amount');
         if (totalAmount) {
-            totalAmount.dataset.total = totalAmount.dataset.total;
+            totalAmount.textContent = totalAmount.textContent;
```
There's a lot of "Hack" comments in this file 😬

glaubersilva commented:
@jonwaldstein I implemented all suggested changes, let me know if there is anything more! ;)

jonwaldstein reviewed (approved):
@glaubersilva thanks for the updates! please send this off to Q/A for final approval

rickalday reviewed (approved):
Tested with Legacy, MSF, and Classic templates.
Working with Donor Opt-in  and Forced Opt-in options too.
